sumpy/e2e.py

```diff
@@ -656,54 +651,33 @@ class M2LPreprocessMultipole(E2EBase):
 
     default_name = "m2l_preprocess_multipole"
 
-    def get_loopy_insns(self, result_dtype):
-        src_coeff_exprs = [
-            sym.Symbol(f"src_coeff{i}")
-            for i in range(len(self.src_expansion))]
-
-        src_rscale = sym.Symbol("src_rscale")
-
-        from sumpy.assignment_collection import SymbolicAssignmentCollection
-        sac = SymbolicAssignmentCollection()
-
-        preprocessed_src_coeff_names = [
-                sac.assign_unique(f"preprocessed_src_coeff{i}", coeff_i)
-                for i, coeff_i in enumerate(
-                    self.tgt_expansion.m2l_translation.preprocess_multipole_exprs(
-                        self.tgt_expansion, self.src_expansion, src_coeff_exprs,
-                        sac=sac, src_rscale=src_rscale))]
-
-        sac.run_global_cse()
-
-        from sumpy.codegen import to_loopy_insns
-        return to_loopy_insns(
-                sac.assignments.items(),
-                vector_names={"d"},
-                pymbolic_expr_maps=[self.tgt_expansion.get_code_transformer()],
-                retain_names=preprocessed_src_coeff_names,
-                complex_dtype=to_complex_dtype(result_dtype),
-                )
-
     def get_kernel(self, result_dtype):
+        m2l_translation = self.tgt_expansion.m2l_translation
         nsrc_coeffs = len(self.src_expansion)
         npreprocessed_src_coeffs = \
-            self.tgt_expansion.m2l_translation.preprocess_multipole_nexprs(
-                self.tgt_expansion, self.src_expansion)
+            m2l_translation.preprocess_multipole_nexprs(self.tgt_expansion,
+                self.src_expansion)
+        child_knl = m2l_translation.preprocess_multipole_loopy_knl(
+            self.tgt_expansion, self.src_expansion, result_dtype)
+
         from sumpy.tools import gather_loopy_arguments
         loopy_knl = lp.make_kernel(
                 [
                     "{[isrc_box]: 0<=isrc_box<nsrc_boxes}",
-                    ],
+                    "{[isrc_coeff]: 0<=isrc_coeff<nsrc_coeffs}",
+                    "{[itgt_coeff]: 0<=itgt_coeff<npreprocessed_src_coeffs}",
+                ],
                 ["""
                 for isrc_box
-                """] + ["""
-                    <> src_coeff{idx} = src_expansions[isrc_box, {idx}]
-                """.format(idx=i) for i in range(nsrc_coeffs)] + [
-                ] + self.get_loopy_insns(result_dtype) + ["""
-                    preprocessed_src_expansions[isrc_box, {idx}] = \
-                        preprocessed_src_coeff{idx}
-                    """.format(idx=i) for i in range(
-                        npreprocessed_src_coeffs)] + ["""
+                    <> coeffs[itgt_coeff] = 0 {id=init, dup=itgt_coeff}
```

Comment:
Use a better name for `coeffs`.

Reply:
Fixed in 25dd7fc4424695f26997e4bf2de55b7f1c0a25c2